File: ios/Classes/internal/ZegoPlatformViewFactory.m
//
//  ZegoPlatformViewFactory.m
//  Pods-Runner
//
//  Created by lizhanpeng@ZEGO on 2020/3/26.
//  Copyright © 2020 Zego. All rights reserved.
//

#import "ZegoPlatformViewFactory.h"
#import "ZegoLog.h"

@interface ZegoPlatformViewFactory()

@property (nonatomic, strong) NSMutableDictionary<NSNumber*, ZegoPlatformView*> *platformViewMap;

@end

@implementation ZegoPlatformViewFactory

+ (instancetype)sharedInstance {
    static ZegoPlatformViewFactory *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[ZegoPlatformViewFactory alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _platformViewMap = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (BOOL)destroyPlatformView:(NSNumber *)viewID {

    ZegoPlatformView *platformView = self.platformViewMap[viewID];

    if (!platformView) {
        ZGLog(@"[destroyPlatformView] platform view for viewID:%d not exists", viewID.intValue);
        [self logCurrentPlatformViews];
        return NO;
    }

    ZGLog(@"[destroyPlatformView] viewID:%d, UIView:%p", viewID.intValue, platformView.view);

    [self.platformViewMap removeObjectForKey:viewID];

    [self logCurrentPlatformViews];

    return YES;
}

- (nullable ZegoPlatformView *)getPlatformView:(NSNumber *)viewID {

    ZGLog(@"[getPlatformView] viewID:%d", viewID.intValue);

    [self logCurrentPlatformViews];
    
    return [self.platformViewMap objectForKey:viewID];
}

- (void)addPlatformView:(ZegoPlatformView *)view viewID:(NSNumber *)viewID {

    ZGLog(@"[createPlatformView] viewID:%d, UIView:%p", viewID.intValue, view.view);

    [self.platformViewMap setObject:view forKey:viewID];

    [self logCurrentPlatformViews];
}

- (void)logCurrentPlatformViews {
    NSMutableString *desc = [[NSMutableString alloc] init];
    for (NSNumber *i in self.platformViewMap) {
        ZegoPlatformView *eachPlatformView = self.platformViewMap[i];
        if (eachPlatformView) {
            [desc appendFormat:@"[ID:%d|View:%p] ", i.intValue, eachPlatformView.view];
        }
    }
    ZGLog(@"[ZegoPlatformViewFactory] currentPlatformViews: %@", desc);
}

#pragma mark FlutterPlatformViewFactory Delegate

#if TARGET_OS_IPHONE
/// Called when dart invoke `createPlatformView`, that is, when Widget `UiKitView` is added to the flutter render tree
- (NSObject<FlutterPlatformView>*)createWithFrame:(CGRect)frame
                                   viewIdentifier:(int64_t)viewId
                                        arguments:(id _Nullable)args {
    
    ZegoPlatformView *view = [[ZegoPlatformView alloc] initWithRect:frame viewID:viewId];
    [self addPlatformView:view viewID:@(viewId)];
    
    return view;
}

#elif TARGET_OS_OSX

- (nonnull NSView *)createWithViewIdentifier:(int64_t)viewId arguments:(nullable id)args {
    ZegoPlatformView *view = [[ZegoPlatformView alloc] initWithRect:CGRectZero viewID:viewId];
    [self addPlatformView:view viewID:@(viewId)];
    
    return view.view;
}

#endif

@end

File: android/src/main/java/im/zego/zego_express_engine/internal/ZegoPlatformViewFactory.java
//
//  android
//  im.zego.zego_express_engine
//
//  Created by Patrick Fu on 2020-04-03.
//  Copyright © 2020 Zego. All rights reserved.
//

package im.zego.zego_express_engine.internal;

import android.content.Context;

import java.util.HashMap;
import java.util.Locale;

import io.flutter.plugin.common.MessageCodec;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;

public class ZegoPlatformViewFactory extends PlatformViewFactory {

    private volatile static ZegoPlatformViewFactory instance;

    private final HashMap<Integer, ZegoPlatformView> platformViews;

    private ZegoPlatformViewFactory(MessageCodec<Object> createArgsCodec) {
        super(createArgsCodec);
        this.platformViews = new HashMap<>();
    }

    public static ZegoPlatformViewFactory getInstance() {
        if (instance == null) {
            synchronized (ZegoPlatformViewFactory.class) {
                if (instance == null) {
                    instance = new ZegoPlatformViewFactory(StandardMessageCodec.INSTANCE);
                }
            }
        }
        return instance;
    }

    /// Called when dart invoke `destroyPlatformView`
    Boolean destroyPlatformView(int viewID) {

        ZegoPlatformView platformView = this.platformViews.get(viewID);

        if (platformView == null) {
            ZegoLog.log("[destroyPlatformView] platform view for viewID:%d not exists", viewID);
            logCurrentPlatformViews();
            return false;
        }

        ZegoLog.log("[destroyPlatformView] viewID:%d, surfaceView:%s", viewID, platformView.getSurfaceView() == null ? "null" : platformView.getSurfaceView().hashCode());

        this.platformViews.remove(viewID);

        logCurrentPlatformViews();

        return true;
    }

    /// Get PlatformView to pass to native when dart invoke `startPreview` or `startPlayingStream`
    ZegoPlatformView getPlatformView(int viewID) {

        ZegoLog.log("[getPlatformView] viewID:%d", viewID);

        logCurrentPlatformViews();

        return this.platformViews.get(viewID);
    }


    private void addPlatformView(int viewID, ZegoPlatformView view) {

        ZegoLog.log("[createPlatformView] viewID:%d, surfaceView:%s", viewID, view.getSurfaceView().hashCode());

        this.platformViews.put(viewID, view);

        logCurrentPlatformViews();
    }

    @Override
    public PlatformView create(Context context, int viewId, Object args) {
        ZegoPlatformView view = new ZegoPlatformView(context);
        this.addPlatformView(viewId, view);
        return view;
    }

    private void logCurrentPlatformViews() {
        StringBuilder desc = new StringBuilder();
        for (Integer id: this.platformViews.keySet()) {
            ZegoPlatformView eachPlatformView = this.platformViews.get(id);
            if (eachPlatformView == null) {
                continue;
            }
            desc.append(String.format(Locale.ENGLISH, "[ID:%d|View:%s] ", id, eachPlatformView.getSurfaceView() == null ? "null" : eachPlatformView.getSurfaceView().hashCode()));
        }
        ZegoLog.log("[ZegoPlatformViewFactory] currentPlatformViews: %s", desc.toString());
    }
}
